### PR TITLE
Make _swift_willThrow atomic

### LINKED ISSUE #62348: `_swift_willThrow` is not atomic.
**Description**

The runtime exports a symbol, `_swift_willThrow`, for use by testing frameworks (specifically XCTest) to track when errors are thrown for diagnostic purposes. This symbol is not declared as an atomic value, which means that access to it has undefined behavior across threads. Because test logic can run concurrently (and that includes throwing errors), atomic access to this variable is necessary.

**Steps to reproduce**

N/A

**Expected behavior**

N/A

### PR DESCRIPTION
**Pull request by grynspan**
This change makes `_swift_willThrow` a C++ atomic value instead of a bare pointer. This change is necessary because the pointer can be loaded across multiple threads concurrently.

Resolves #62348.

### CODE REVIEW

**Review by mikeash — APPROVED** [review]

**Review by stmontgomery — APPROVED** [review]

**Comment by grynspan** [comment]
@swift-ci please test

**Review by mikeash — APPROVED** [review]

File: stdlib/public/runtime/ErrorObjectTestSupport.h [hunk 1]
```diff
@@ -17,10 +17,20 @@
...
 #define SWIFT_RUNTIME_ERROROBJECT_TEST_SUPPORT_H
 
+#include <atomic>
+
 namespace swift {
 
-SWIFT_RUNTIME_EXPORT void (*_swift_willThrow)(SwiftError *error);
+#if defined(__cplusplus)
+SWIFT_RUNTIME_EXPORT std::atomic<void (*)(SwiftError *error)> _swift_willThrow;
+#endif
```
[the remark this answers is not in the document]
Comment (grynspan) [reply stdlib/public/runtime/ErrorObjectTestSupport.h:26]: There's no ODR violation as this is the only definition of the variable. Is there another one somewhere in the repo I missed?

This should not be an MVP as there's no cast nor function call here. It's a plain old variable declaration. Type is `std::atomic<void (*)(SwiftError *error)>`, i.e. an atomic function pointer.

`SWIFT_RUNTIME_EXPORT` includes `extern`.

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test

**Review by al45tair — APPROVED** [review]
LGTM.

Makes me wonder whether we do this for other callbacks, or whether we're relying on callers using them _before_ going multithreaded or inserting their own barriers as appropriate.

**Comment by mikeash** [comment]
We have hooks for retain/release and for the default global executor. THey're all implicitly relying on being set before multithreaded access happens. Retain/release is in practice set early in a static initializer inserted with `DYLD_INSERT_LIBRARIES`. The global executor hooks are up to whatever might use them, there's no specific use case there. We should probably standardize everything on something sensible like this.

**Comment by grynspan** [comment]
> We have hooks for retain/release and for the default global executor. THey're all implicitly relying on being set before multithreaded access happens. Retain/release is in practice set early in a static initializer inserted with `DYLD_INSERT_LIBRARIES`. The global executor hooks are up to whatever might use them, there's no specific use case there. We should probably standardize everything on something sensible like this.

The difference with this callback is, I think, that it's *known* that we set it late in execution, in another module, after other threads or tasks have already been spun up. That's why I opted to change it. I don't personally want to go around messing with those other callbacks unless somebody can show an example of them being set later in execution.

**Comment by grynspan** [comment]
FWIW the overhead of making the global executor hook atomic is going to be negligible, so if it can be set late or arbitrarily, I'd agree we should take steps to make it safer.

I wouldn't dare say the overhead is negligible for retain/release, though.

**Pull request merged**